### DISCUSSION TITLE
[Deployment optimisation] Generating static pages while running the server

### DISCUSSION THREAD

**Discussion by iKonrad** [vercel/next.js#42459] [Ideas]
### Describe the feature you'd like to request

I think it's more of a request for open discussion since I'm not super familiar of the processes behind the static page generation logic. 

We can all agree that nothing can beat the performance of serving statically generated pages, and NextJS provides 2 core ways to serve them:

- By running NextJS server
- By exporting html files and uploading them to some CDN.

Both ways have pros and cons, but they both share the same disadvantage comparing to SSR pages: extremely long deployment time. And worst part is that the deployment time increases along with the number of generated pages.

I've got project with 30k+ pages, and if not some CI/CD hacks with removing folders from `pages` folder (for pages we dont need to rebuild every time), the total build + export time easily can easily take over 1 hour.

So, to serve pages statically with NextJS server, you still need to prebuild all pages during the deployment. OK, I get that we don't need to generate all pages and let them to build during the first visit, but that approach is sort of against the whole idea of generating pages in the first place (especially for SEO purposes since if Googlebot will be the first one to visit the url, there's no speed improvement at all).

### Describe the solution you'd like

What if there was an option to run a NextJS server, where all SSG pages would behave like 'blocking' type, and just after the server is boot up, pages could keep being generated while still accepting traffic? So if there's a visit before the page was generated, it would generate upon visit, if not, it would generate during the initial build.

So, TLDR, would it make sense, and would it be possible at all considering the NextJS infrastructure, to run the NextJS server ASAP, and run the SSG pages generation in the background?

### Describe alternatives you've considered

So far to speed up the deployment time (while still prebuilding all pages when we need), were some hacks with removing folders from the `pages` folder to generate only sets of pages that we need. Not perfect, but this only works when HTML files are exported, won't work with NextJS server.
